libpam/pam_item.c

```diff
@@ -38,6 +38,12 @@ int pam_set_item (pam_handle_t *pamh, int item_type, const void *item)
     switch (item_type) {
 
     case PAM_SERVICE:
+	if (item == NULL) {
+	    pam_syslog(pamh, LOG_ERR,
+		       "pam_set_item: attempt to set service to NULL");
+	    retval = PAM_BAD_ITEM;
+	    break;
+	}
 	/* Setting handlers_loaded to 0 will cause the handlers
```

Comment:
Given that without this change pam_set_item(pamh, PAM_SERVICE, NULL) just segfaults, and pam_start doesn't allow NULL service name, setting service to NULL is essentially not allowed anyway, so I think it's safe to apply this change.

With regards to the return value, PAM_BAD_ITEM is probably OK: it's the closest error code among those that are listed in the documentation for pam_set_item.  Apparently, pam_set_item(pamh, PAM_CONV, NULL) returns PAM_PERM_DENIED, but this is not documented.